File: python-sdk/src/graphmind_context_graphs/db/schema.py
"""Bootstrap schema indexes in Graphmind."""

from __future__ import annotations

import logging
from typing import Any

from .queries import SCHEMA_QUERIES, create_vector_index


def bootstrap_schema(
    client: Any,
    graph: str,
    dimensions: int,
    metric: str,
    logger: logging.Logger,
) -> None:
    logger.info("Bootstrapping schema for graph: %s", graph)

    # Property indexes
    for name, query in SCHEMA_QUERIES.items():
        try:
            client.query(query, graph=graph)
        except Exception as e:
            logger.debug("Index may already exist: %s", e)

    # Vector indexes
    vector_indexes = [
        ("intent_embedding", "Intent"),
        ("trace_embedding", "DecisionTrace"),
        ("concept_embedding", "Concept"),
    ]
    for idx_name, label in vector_indexes:
        try:
            client.query(create_vector_index(idx_name, label, dimensions, metric), graph=graph)
        except Exception as e:
            logger.debug("Vector index may already exist: %s", e)

    logger.info("Schema bootstrap complete")

File: src/db/schema.ts
import type { GraphmindClient } from "graphmind-sdk";
import type { Logger } from "../utils/logger.js";
import { SCHEMA_QUERIES } from "./queries.js";

export async function bootstrapSchema(
  client: GraphmindClient,
  graph: string,
  dimensions: number,
  metric: string,
  logger: Logger
): Promise<void> {
  logger.info("Bootstrapping schema for graph: %s", graph);

  // Create property indexes
  const indexQueries = [
    SCHEMA_QUERIES.createIntentIndex,
    SCHEMA_QUERIES.createConstraintIndex,
    SCHEMA_QUERIES.createActionIndex,
    SCHEMA_QUERIES.createTraceIndex,
    SCHEMA_QUERIES.createProjectIndex,
    SCHEMA_QUERIES.createDomainIndex,
    SCHEMA_QUERIES.createConceptIndex,
    SCHEMA_QUERIES.createToolIndex,
    SCHEMA_QUERIES.createAgentIndex,
    SCHEMA_QUERIES.createSkillIndex,
  ];

  for (const query of indexQueries) {
    try {
      await client.query(query, graph);
    } catch (err) {
      logger.debug("Index may already exist: %s", (err as Error).message);
    }
  }

  // Create vector indexes
  const vectorQueries = [
    SCHEMA_QUERIES.createIntentVectorIndex(dimensions, metric),
    SCHEMA_QUERIES.createTraceVectorIndex(dimensions, metric),
    SCHEMA_QUERIES.createConceptVectorIndex(dimensions, metric),
  ];

  for (const query of vectorQueries) {
    try {
      await client.query(query, graph);
    } catch (err) {
      logger.debug(
        "Vector index may already exist: %s",
        (err as Error).message
      );
    }
  }

  logger.info("Schema bootstrap complete");
}
